src/Data/Persist/Internals.hs

```diff
@@ -0,0 +1,225 @@
+  {-# INLINE pure #-}
+
+  f <*> a = Get $ \e p -> do
+    p' :!: f' <- unGet f e p
+    p'' :!: a' <- unGet a e p'
+    pure $! p'' :!: f' a'
+  {-# INLINE (<*>) #-}
+
+  m1 *> m2 = do
+    void m1
+    m2
+  {-# INLINE (*>) #-}
+
+instance Monad Get where
+  m >>= f = Get $ \e p -> do
+    p' :!: x <- unGet m e p
+    unGet (f x) e p'
+  {-# INLINE (>>=) #-}
+
+  fail = Fail.fail
+  {-# INLINE fail #-}
+
+data GetException =
+    LengthException Int String
+  | CharException Int String
+  | EOFException Int String
+  | GenericGetException Int String
+  deriving (Eq, Show)
+
+instance Exception GetException
+
+instance Fail.MonadFail Get where
+  fail msg = getFail GenericGetException ("Failed reading: " <> msg)
+  {-# INLINE fail #-}
+
+getOffset :: Get Int
+getOffset = Get $ \e p -> pure $! p :!: (p `minusPtr` (geBegin e))
+{-# INLINE getOffset #-}
+
+getFail :: (Int -> String -> GetException) -> String -> Get a
+getFail ctor msg = do
+  offset <- getOffset
+  Get $ \_ _ -> throwIO (ctor offset msg)
+
+runGetIO :: Get a -> ByteString -> IO a
+runGetIO m s = run
+  where run = withForeignPtr buf $ \p -> allocaBytes 8 $ \t -> do
+          let env = GetEnv { geBuf = buf, geBegin = p, geEnd = p `plusPtr` (pos + len), geTmp = t }
+          _ :!: r <- unGet m env (p `plusPtr` pos)
+          pure r
+        (B.PS buf pos len) = s
+
+-- | Run the Get monad applies a 'get'-based parser on the input ByteString
+runGet :: Get a -> ByteString -> Either String a
+runGet m s = unsafePerformIO $ catch (run (runGetIO m s)) handler
+  where run mx = do
+          x <- mx
+          pure $! Right x
+        handler (e :: GetException) = pure $ Left $ displayException e
+{-# NOINLINE runGet #-}
+
+data Chunk = Chunk
+  { chkBegin :: !(Ptr Word8)
+  , chkEnd   :: !(Ptr Word8)
+  }
+
+data PutEnv = PutEnv
+  { peChks :: !(IORef (NonEmpty Chunk))
+  , peEnd  :: !(IORef (Ptr Word8))
+  , peTmp  :: !(Ptr Word8)
+  }
+
+newtype Put a = Put
+  { unPut :: PutEnv -> Ptr Word8 -> IO (Pair (Ptr Word8) a) }
+
+instance Functor Put where
+  fmap f m = Put $ \e p -> do
+    p' :!: x <- unPut m e p
+    pure $! p' :!: f x
+  {-# INLINE fmap #-}
+
+instance Applicative Put where
+  pure a = Put $ \_ p -> pure $! p :!: a
+  {-# INLINE pure #-}
+
+  f <*> a = Put $ \e p -> do
+    p' :!: f' <- unPut f e p
+    p'' :!: a' <- unPut a e p'
+    pure $! p'' :!: f' a'
+  {-# INLINE (<*>) #-}
+
+  m1 *> m2 = do
+    void m1
+    m2
+  {-# INLINE (*>) #-}
+
+instance Monad Put where
+  m >>= f = Put $ \e p -> do
+    p' :!: x <- unPut m e p
+    unPut (f x) e p'
+  {-# INLINE (>>=) #-}
+
+minChunkSize :: Int
+minChunkSize = 0x10000
+{-# INLINE minChunkSize #-}
+
+newChunk :: Int -> IO Chunk
+newChunk size = do
+  let n = max size minChunkSize
+  p <- mallocBytes n
+  pure $! Chunk p $ p `plusPtr` n
+{-# INLINE newChunk #-}
+
+doGrow :: PutEnv -> Ptr Word8 -> Int -> IO (Pair (Ptr Word8) ())
+doGrow e p n = do
+  k <- newChunk n
+  modifyIORef' (peChks e) $ \case
```

Comment:
Maybe doGrow needs a better name. Or don't expose it. This function is truly internal. It is only separate because I wanted this noinline pragma. It could also be where bound within grow.

Reply:
I pushed grow into the Internals module and exposed it, then hid doGrow.